### PR TITLE
r/sagemaker_feature_group - new resource

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_sagemaker_feature_group - add new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerFeatureGroup_'
--- PASS: TestAccAWSSagemakerFeatureGroup_disappears (49.09s)
--- PASS: TestAccAWSSagemakerFeatureGroup_description (54.42s)
--- PASS: TestAccAWSSagemakerFeatureGroup_multipleFeatures (60.96s)
--- PASS: TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig (63.50s)
--- PASS: TestAccAWSSagemakerFeatureGroup_basic (64.97s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog (84.52s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog (87.71s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (93.84s)
--- PASS: TestAccAWSSagemakerFeatureGroup_tags (140.86s)
```

### CODE REVIEW

File: aws/resource_aws_sagemaker_feature_group.go [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+	arn := aws.StringValue(output.FeatureGroupArn)
+	d.Set("feature_group_name", output.FeatureGroupName)
+	d.Set("event_time_feature_name", output.EventTimeFeatureName)
+	d.Set("description", output.Description)
+	d.Set("record_identifier_feature_name", output.RecordIdentifierFeatureName)
+	d.Set("role_arn", output.RoleArn)
+	d.Set("arn", arn)
+
+	if err := d.Set("feature_definition", flattenSagemakerFeatureGroupFeatureDefinition(output.FeatureDefinitions)); err != nil {
+		return fmt.Errorf("error setting feature_definition for Sagemaker Feature Group (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_feature_group.go:262]:
```suggestion
		return fmt.Errorf("error setting feature_definition for SageMaker Feature Group (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_feature_group.go [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+	}
+
+	if err := d.Set("online_store_config", flattenSagemakerFeatureGroupOnlineStoreConfig(output.OnlineStoreConfig)); err != nil {
+		return fmt.Errorf("error setting online_store_config for Sagemaker Feature Group (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_feature_group.go:266]:
```suggestion
		return fmt.Errorf("error setting online_store_config for SageMaker Feature Group (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_feature_group.go [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+	}
+
+	if err := d.Set("offline_store_config", flattenSagemakerFeatureGroupOfflineStoreConfig(output.OfflineStoreConfig)); err != nil {
+		return fmt.Errorf("error setting offline_store_config for Sagemaker Feature Group (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_feature_group.go:270]:
```suggestion
		return fmt.Errorf("error setting offline_store_config for SageMaker Feature Group (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_feature_group.go [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+	}
+
+	tags, err := keyvaluetags.SagemakerListTags(conn, arn)
+	if err != nil {
+		return fmt.Errorf("error listing tags for Sagemaker Feature Group (%s): %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_feature_group.go:275]:
```suggestion
		return fmt.Errorf("error listing tags for SageMaker Feature Group (%s): %w", d.Id(), err)
```

File: aws/resource_aws_sagemaker_feature_group.go [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+}
+
+func resourceAwsSagemakerFeatureGroupUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+
+		if err := keyvaluetags.SagemakerUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating Sagemaker Feature Store (%s) tags: %w", d.Id(), err)
```
Comment (ewbankkit) [inline aws/resource_aws_sagemaker_feature_group.go:292]:
```suggestion
			return fmt.Errorf("error updating SageMaker Feature Group (%s) tags: %w", d.Id(), err)
```

**Comment by ewbankkit** [comment]
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerFeatureGroup_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerFeatureGroup_ -timeout 120m
=== RUN   TestAccAWSSagemakerFeatureGroup_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_description
=== PAUSE TestAccAWSSagemakerFeatureGroup_description
=== RUN   TestAccAWSSagemakerFeatureGroup_tags
=== PAUSE TestAccAWSSagemakerFeatureGroup_tags
=== RUN   TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== PAUSE TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== RUN   TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== PAUSE TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_disappears
=== PAUSE TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== CONT  TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== CONT  TestAccAWSSagemakerFeatureGroup_tags
=== CONT  TestAccAWSSagemakerFeatureGroup_description
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
    resource_aws_sagemaker_feature_group_test.go:267: Step 1/2 error: Error running apply: 2021/01/05 17:18:52 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.12.0
        
        Error: error creating SageMaker Feature Group: ValidationException: The execution role ARN is invalid. Please ensure that the role exists and that its trust relationship policy allows the action 'sts:AssumeRole' for the service principal 'sagemaker.amazonaws.com'.
        	status code: 400, request id: 650c3e63-04f5-4cb8-8d3e-093e4006b4e6
        
        
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
    resource_aws_sagemaker_feature_group_test.go:303: Step 1/2 error: Error running apply: 2021/01/05 17:18:51 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.12.0
        
        Error: error creating SageMaker Feature Group: ValidationException: The execution role ARN is invalid. Please ensure that the role exists and that its trust relationship policy allows the action 'sts:AssumeRole' for the service principal 'sagemaker.amazonaws.com'.
        	status code: 400, request id: 76c5898e-c8ab-4dce-86e5-6dba370577c0
        
        
--- FAIL: TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog (23.37s)
--- FAIL: TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog (24.20s)
--- PASS: TestAccAWSSagemakerFeatureGroup_multipleFeatures (41.02s)
--- PASS: TestAccAWSSagemakerFeatureGroup_description (41.53s)
--- PASS: TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig (55.38s)
--- PASS: TestAccAWSSagemakerFeatureGroup_disappears (58.55s)
--- PASS: TestAccAWSSagemakerFeatureGroup_tags (64.60s)
--- PASS: TestAccAWSSagemakerFeatureGroup_basic (65.25s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (72.15s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	72.262s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

and then

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerFeatureGroup_offlineConfig_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerFeatureGroup_offlineConfig_basic -timeout 120m
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
    resource_aws_sagemaker_feature_group_test.go:236: Step 1/2 error: Error running apply: 2021/01/05 17:30:48 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.12.0
        
        Error: error creating SageMaker Feature Group: ValidationException: The execution role ARN is invalid. Please ensure that the role exists and that its trust relationship policy allows the action 'sts:AssumeRole' for the service principal 'sagemaker.amazonaws.com'.
        	status code: 400, request id: c6aa741c-6c7e-4bde-96e8-e357f653dea8
        
        
--- FAIL: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (12.55s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	12.646s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

which passed in a previous run.
This suggests an IAM eventual consistency issue: https://github.com/hashicorp/terraform-provider-aws/issues/16752.

**Comment by DrFaust92** [comment]
Added retry for IAM error mentioned above

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerFeatureGroup_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerFeatureGroup_ -timeout 120m
=== RUN   TestAccAWSSagemakerFeatureGroup_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_description
=== PAUSE TestAccAWSSagemakerFeatureGroup_description
=== RUN   TestAccAWSSagemakerFeatureGroup_tags
=== PAUSE TestAccAWSSagemakerFeatureGroup_tags
=== RUN   TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== PAUSE TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== RUN   TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== PAUSE TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_disappears
=== PAUSE TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== CONT  TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== CONT  TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== CONT  TestAccAWSSagemakerFeatureGroup_tags
=== CONT  TestAccAWSSagemakerFeatureGroup_description
--- PASS: TestAccAWSSagemakerFeatureGroup_disappears (39.49s)
--- PASS: TestAccAWSSagemakerFeatureGroup_multipleFeatures (49.82s)
--- PASS: TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig (50.46s)
--- PASS: TestAccAWSSagemakerFeatureGroup_description (56.64s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog (58.49s)
--- PASS: TestAccAWSSagemakerFeatureGroup_basic (58.53s)
--- PASS: TestAccAWSSagemakerFeatureGroup_tags (59.22s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (68.68s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog (68.84s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	68.949s
```

**Comment by DrFaust92** [comment]
Added retry for S3 access denied issue and descoped s3 test permissions:

```
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog (99.63s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (109.86s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog (121.75s)
```

**Comment by breathingdust** [comment]
LGTM :rocket: Thanks @DrFaust92!

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerFeatureGroup_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerFeatureGroup_ -timeout 120m
=== RUN   TestAccAWSSagemakerFeatureGroup_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_description
=== PAUSE TestAccAWSSagemakerFeatureGroup_description
=== RUN   TestAccAWSSagemakerFeatureGroup_tags
=== PAUSE TestAccAWSSagemakerFeatureGroup_tags
=== RUN   TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== PAUSE TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== RUN   TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== PAUSE TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== PAUSE TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
=== RUN   TestAccAWSSagemakerFeatureGroup_disappears
=== PAUSE TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_basic
=== CONT  TestAccAWSSagemakerFeatureGroup_tags
=== CONT  TestAccAWSSagemakerFeatureGroup_multipleFeatures
=== CONT  TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig
=== CONT  TestAccAWSSagemakerFeatureGroup_disappears
=== CONT  TestAccAWSSagemakerFeatureGroup_description
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog
=== CONT  TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog
--- PASS: TestAccAWSSagemakerFeatureGroup_multipleFeatures (34.67s)
--- PASS: TestAccAWSSagemakerFeatureGroup_basic (34.71s)
--- PASS: TestAccAWSSagemakerFeatureGroup_onlineConfigSecurityConfig (38.67s)
--- PASS: TestAccAWSSagemakerFeatureGroup_disappears (47.89s)
--- PASS: TestAccAWSSagemakerFeatureGroup_description (48.45s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_createCatalog (49.26s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_providedCatalog (52.14s)
--- PASS: TestAccAWSSagemakerFeatureGroup_offlineConfig_basic (63.03s)
--- PASS: TestAccAWSSagemakerFeatureGroup_tags (65.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	66.697s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

As this resource covers a component of SageMaker Studio it is not available in GovCloud. The tests currently fail with the following. An issue to cover allowing these tests to skip will be created and linked back to this issue.

```
=== CONT  TestAccAWSSagemakerFeatureGroup_basic
    resource_aws_sagemaker_feature_group_test.go:66: Step 1/2 error: Error running apply: 2021/01/18 09:07:49 [DEBUG] Using modified User-Agent: Terraform/0.12.29 HashiCorp-terraform-exec/0.12.0

        Error: Error creating SageMaker Feature Group: UnknownOperationException: The requested operation is not supported in the called region.
        	status code: 400, request id: 24cef7ef-5040-484b-bdd5-829420ec9ea1


```

**Pull request merged**